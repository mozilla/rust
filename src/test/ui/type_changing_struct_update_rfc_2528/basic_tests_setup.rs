// gate-test-type_changing_struct_update

/// Here, we're documenting the current state of affairs related to
/// FRU (functional record update)

#![feature(type_changing_struct_update)]

struct Machine<S> {
    state: S,
    common_field1: &'static str,
    common_field2: i32,
}

struct State1;
struct State2;

impl Machine<State1> {
    fn into_state2(self) -> Machine<State2> {
        // do stuff
        Machine {
            state: State2,
            ..self
        }
    }
}

#[test]
#[should_panic]  // NOTE: this would be fixed by the implementation of the RFC
fn update_to_state2() {
    let m1: Machine<State1> = Machine {
        state: State1,
        common_field1: "hello",
        common_field2: 2,
    };
    let m2: Machine<State2> = m1.into_state2();
    assert_eq!(State2, m2.state);
}

fn main() {}
