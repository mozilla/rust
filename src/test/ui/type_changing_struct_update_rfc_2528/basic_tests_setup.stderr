error[E0308]: mismatched types
  --> $DIR/basic_tests_setup.rs:19:15
   |
LL |             ..self
   |               ^^^^ expected struct `State2`, found struct `State1`
   |
   = note: expected struct `Machine<State2>`
              found struct `Machine<State1>`
